misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NodeFormattingMetrics.java

```diff
@@ -0,0 +1,116 @@
+            this.grandParent = parent.parent();
+        } else {
+            this.grandParent = null;
+        }
+        setNodeMetrics();
+    }
+
+    private void setNodeMetrics() {
+        if (node instanceof BuiltinSimpleNameReferenceNode) {
+            handleBuiltinSimpleNameReferenceNode();
+        }
+    }
+
+    boolean handleBuiltinSimpleNameReferenceNode() {
+        boolean addSpaces = true;
+        ArrayList<SyntaxKind> parentNodes = new ArrayList<>(
+                Arrays.asList(
+                        SyntaxKind.FUNCTION_CALL,
+                        SyntaxKind.TYPE_TEST_EXPRESSION,
+                        SyntaxKind.TYPE_PARAMETER,
+                        SyntaxKind.TYPE_CAST_PARAM,
+                        SyntaxKind.UNION_TYPE_DESC,
+                        SyntaxKind.XML_TYPE_DESC));
+        if (parent != null && (parentNodes.contains(parent.kind()) || grandParent != null &&
+                ((parent.kind().equals(SyntaxKind.TYPED_BINDING_PATTERN) &&
+                        grandParent.kind().equals(SyntaxKind.FOREACH_STATEMENT)) ||
+                (parent.kind().equals(SyntaxKind.UNION_TYPE_DESC) &&
+                        grandParent.kind().equals(SyntaxKind.PARAMETERIZED_TYPE_DESC)) ||
+                (parent.kind().equals(SyntaxKind.OBJECT_FIELD) &&
+                        ((ObjectFieldNode) parent).visibilityQualifier().isPresent()) ||
+                (grandParent.kind().equals(SyntaxKind.LOCAL_VAR_DECL) &&
+                        grandParent.children().get(1).equals(parent)) ||
+                grandParent.kind().equals(SyntaxKind.FROM_CLAUSE) ||
+                grandParent.kind().equals(SyntaxKind.PARAMETERIZED_TYPE_DESC)))) {
+            addSpaces = false;
+        }
+        if (parent != null && grandParent != null && (parent.kind().equals(SyntaxKind.UNION_TYPE_DESC) &&
+                !grandParent.kind().equals(SyntaxKind.TYPE_CAST_PARAM) && parent.children().get(0).equals(node))) {
+            addSpaces = true;
+        }
+        return addSpaces;
+    }
+
+    boolean handleSimpleNameReferenceNode() {
+        boolean addSpaces = false;
+        ArrayList<SyntaxKind> parentNodes = new ArrayList<>(
+                Arrays.asList(
+                        SyntaxKind.TYPED_BINDING_PATTERN,
+                        SyntaxKind.FIELD_ACCESS,
+                        SyntaxKind.RECORD_FIELD_WITH_DEFAULT_VALUE,
+                        SyntaxKind.OPTIONAL_TYPE_DESC,
+                        SyntaxKind.ARRAY_TYPE_DESC));
+        if (parent != null && ((parentNodes.contains(parent.kind())) || grandParent != null &&
+                ((parent.kind().equals(SyntaxKind.INDEXED_EXPRESSION) &&
+                        grandParent.kind().equals(SyntaxKind.ASSIGNMENT_STATEMENT)) ||
+                ((parent.kind().equals(SyntaxKind.ASYNC_SEND_ACTION) ||
+                parent.kind().equals(SyntaxKind.ASSIGNMENT_STATEMENT)) &&
+                        parent.children().get(0).equals(node))))) {
```

Comment:
Same as above

Reply:
Planning on refactoring and simplifying the logic separately in a different PR.